pkg/validators/quota.go

```diff
@@ -86,149 +88,150 @@ func validateResourceRequirements(rs []ResourceRequirement, up *usageProvider) (
+		qe  []QuotaError
+		err error
 	}
-
-	return qe, nil
-}
-
-func validateServiceLimits(rs []ResourceRequirement, ls []*sub.ConsumerQuotaMetric, up *usageProvider) ([]QuotaError, error) {
-	// Group by Metric and Aggregation
-	type gk struct {
-		Metric      string
-		Aggregation string
+	ch := make(chan chs)
+	for k, g := range groups { // Spawn
+		go func(k gk, g []ResourceRequirement) {
+			qe, err := validateServiceRequirements(k.Consumer, k.Service, g, up)
+			ch <- chs{qe, err}
+		}(k, g)
 	}
-	groups := map[gk][]ResourceRequirement{}
-	for _, r := range rs {
-		k := gk{r.Metric, r.Aggregation}
-		groups[k] = append(groups[k], r)
+	errs := config.Errors{}
+	qerrs := []QuotaError{}
+	for range groups { // Gather
+		s := <-ch
+		qerrs = append(qerrs, s.qe...)
+		errs.Add(s.err)
 	}
+	return qerrs, errs.OrNil()
+}
 
-	qe := []QuotaError{}
-	for k, g := range groups {
-		agg, err := aggregation(k.Aggregation)
-		if err != nil {
-			return qe, err
-		}
+// Validate requirements for a single consume & service pair.
+func validateServiceRequirements(consumer string, service string, rs []ResourceRequirement, up *usageProvider) ([]QuotaError, error) {
+	qms, err := queryMetrics(consumer, service)
+	if err != nil {
+		return nil, err
+	}
 
-		// select limits for the metric
-		ml := []*sub.ConsumerQuotaLimit{}
-		for _, l := range ls {
-			if l.Metric == k.Metric {
-				ml = append(ml, l.ConsumerQuotaLimits...)
-			}
-		}
-		if len(ml) == 0 {
-			return qe, fmt.Errorf("limits for metric %q were not found", k.Metric)
-		}
+	errs := config.Errors{}
+	reqToBuckets, err := gatherBucketsRequirements(rs, qms)
+	errs.Add(err)
+	qerrs := []QuotaError{}
+	for _, br := range reqToBuckets {
+		qerrs = append(qerrs, validateBucket(br, up)...)
+	}
+	return qerrs, errs.OrNil()
+}
 
-		for _, limit := range ml {
-			qle := validateLimit(g, limit, up, agg)
-			qe = append(qe, qle...)
+// Find a bucket in the ConsumerQuotaLimit that matches the ResourceRequirement.
+func findBucket(r ResourceRequirement, ql *sub.ConsumerQuotaLimit) (*sub.QuotaBucket, error) {
+	// Iterate buckets in order from most to less specific
+	for i := len(ql.QuotaBuckets) - 1; i >= 0; i-- {
+		if r.InBucket(ql.QuotaBuckets[i]) {
+			return ql.QuotaBuckets[i], nil
 		}
 	}
-	return qe, nil
+	// According to docs the top bucket should be a wildcard `dimensions={}`
+	// So we should never end up here, return fake "unlimited" bucket and report error.
+	return &sub.QuotaBucket{Dimensions: map[string]string{}, EffectiveLimit: -1},
+		fmt.Errorf("unexpected default-less ConsumerQuotaLimit: %q", ql.Name)
 }
 
-func validateLimit(rs []ResourceRequirement, limit *sub.ConsumerQuotaLimit, up *usageProvider, agg aggFn) []QuotaError {
-	qe := []QuotaError{}
-	for _, bucket := range limit.QuotaBuckets {
-		vs := []int64{}
-		for _, r := range rs {
-			if r.InBucket(bucket) {
-				vs = append(vs, r.Required)
-			}
-		}
-		if len(vs) == 0 {
+// Set of requirements to the single bucket with context attached
```

Comment:
I understand this as "The set of requirements that match a specific bucket", emphasis on "match". My assumption is that there has been processing to match the requirements to the bucket. The wording of "requirements to bucket" is confusing. If so please update the description.

Reply:
Changed to `Set of requirements that fall into the specific bucket `